sparse/coo/indexing.py

```diff
@@ -0,0 +1,444 @@
+            shape.append(len(range(ind.start, ind.stop, ind.step)))
+            dt = np.min_scalar_type(min(-(dim - 1) if dim != 0 else -1 for dim in shape))
+            coords.append((x.coords[i, mask].astype(dt) - ind.start) // ind.step)
+            i += 1
+        elif isinstance(ind, Iterable):
+            raise NotImplementedError('Advanced indexing is not yet supported.')
+        # Add a dimension for None.
+        elif ind is None:
+            coords.append(np.zeros(n))
+            shape.append(1)
+
+    # Join all the transformed coords.
+    if coords:
+        coords = np.stack(coords, axis=0)
+    else:
+        # If index result is a scalar, return a 0-d COO or
+        # a scalar depending on whether the last index is an ellipsis.
+        if last_ellipsis:
+            coords = np.empty((0, n), dtype=np.uint8)
+        else:
+            if n != 0:
+                return x.data[mask][0]
+            else:
+                return _zero_of_dtype(x.dtype)[()]
+
+    shape = tuple(shape)
+    data = x.data[mask]
+
+    return COO(coords, data, shape=shape,
+               has_duplicates=x.has_duplicates,
+               sorted=x.sorted)
+
+
+def _mask(coords, indices, shape):
+    indices = _prune_indices(indices, shape)
+
+    ind_ar = np.empty((len(indices), 3), dtype=np.intp)
+
+    for i, idx in enumerate(indices):
+        if isinstance(idx, slice):
+            ind_ar[i] = [idx.start, idx.stop, idx.step]
+        else:  # idx is an integer
+            ind_ar[i] = [idx, idx + 1, 1]
+
+    mask, is_slice = _compute_mask(coords, ind_ar)
+
+    if is_slice:
+        return slice(mask[0], mask[1], 1)
+    else:
+        return mask
+
+
+def _prune_indices(indices, shape, prune_none=True):
+    """
+    Gets rid of the indices that do not contribute to the
+    overall mask, e.g. None and full slices.
+
+    Parameters
+    ----------
+    indices : tuple
+        The indices to the array.
+    shape : tuple[int]
+        The shape of the array.
+
+    Returns
+    -------
+    indices : tuple
+        The filtered indices.
+
+    Examples
+    --------
+    >>> _prune_indices((None, 5), (10,)) # None won't affect the mask
+    [5]
+    >>> _prune_indices((slice(0, 10, 1),), (10,)) # Full slices don't affect the mask
+    []
+    """
+    if prune_none:
+        indices = [idx for idx in indices if idx is not None]
+
+    i = 0
+    for idx, l in zip(indices[::-1], shape[::-1]):
+        if not isinstance(idx, slice):
+            break
+
+        if idx.start == 0 and idx.stop == l and idx.step == 1:
+            i += 1
+            continue
+
+        if idx.start == l - 1 and idx.stop == -1 and idx.step == -1:
+            i += 1
+            continue
+
+        break
+    if i != 0:
+        indices = indices[:-i]
+    return indices
+
+
+@numba.jit(nopython=True)
+def _compute_mask(coords, indices):  # pragma: no cover
+    """
+    Gets the mask for the coords given the indices in slice format.
```

Comment:
We should define slice format somewhere.  Either here or in _mask above (and then refer to that function here).  I think you mean the (start, stop, step) triples.

Reply:
That seems like a good idea.

Reply:
It's already defined in the parameters section.